### PR TITLE
fix: Application.Restart throws InvalidOperationException

### LINKED ISSUE #2769: Application.Restart throws InvalidOperationException
* .NET Core Version: 
3.1.1

* Have you experienced this same bug with .NET Framework?: 
**No - this has always worked fine on .NET Framework!**

**Problem description:**
Calling `Application.Restart` in a .NET Core WinForms application does not work as it does on .NET Framework. An exception is thrown:

```
System.InvalidOperationException: Process was not started by this object, so requested information cannot be determined.
   at System.Diagnostics.Process.get_StartInfo()
   at System.Windows.Forms.Application.Restart()
```

I also tried launching the exe without the debugger, but that doesn't make a difference.

Additionally, there is no [note](https://docs.microsoft.com/en-us/dotnet/api/system.windows.forms.application.restart?view=netcore-3.1) or [breaking change](https://docs.microsoft.com/en-us/dotnet/core/compatibility/fx-core) detailing any .NET Core changes.

**Expected behavior:**
`Application.Restart` should work the same as it does on .NET Framework.

**Minimal repro:**
Create a new .NET Core WinForms project in Visual Studio, put `Application.Restart` anywhere (even under `InitializeComponent` in `Form1`), run it, and observe the exception.

### PR DESCRIPTION
**Pull request by EatonZ**
Fixes #2769

## Proposed changes

Implemented the [suggested fix](https://github.com/dotnet/winforms/issues/2769#issuecomment-579180352) from @stephentoub.

## Customer Impact

This fixes a method broken by changes in .NET Core. Most customers are probably unaware of the broken method since it is not a documented breaking change, nor are there other issues reporting the breakage.

## Regression? 

- Yes

## Risk

There should be no risk. I have tested this fix myself and confirmed that it works.

###### Microsoft Reviewers: [Open in CodeFlow](http://wpcp.azurewebsites.net/CodeFlowProtocolProxy2.php?pullrequest=https://github.com/dotnet/winforms/pull/2845)

### CODE REVIEW

File: src/System.Windows.Forms/src/System/Windows/Forms/Application.cs [hunk 1]
```diff
@@ -1167,7 +1167,7 @@ public static void Restart()
                     sb.Append(arguments[arguments.Length - 1]);
                     sb.Append('"');
                 }
-                ProcessStartInfo currentStartInfo = Process.GetCurrentProcess().StartInfo;
+                ProcessStartInfo currentStartInfo = new ProcessStartInfo();
```
Comment (stephentoub) [inline src/System.Windows.Forms/src/System/Windows/Forms/Application.cs]: Is there a test we can add?

Comment (RussKie) [reply src/System.Windows.Forms/src/System/Windows/Forms/Application.cs]: I don't think this method is unit-testable, there are too many moving parts here. Can't think of a relatively low-effort automation either, we haven't done any automated integration tests yet... 
However I do agree we some form a verification that it is working as intended.

@EatonZ please have a look at https://github.com/dotnet/winforms/pull/2801. I have created few PS scripts that generate net472 and netcoreapp5.0 app, and then ran manual tests against them to confirm the fix.
You could tweak the "single app" script and then run a series of tests with and without input parameters.

A verification would be something along the lines of:
```powershell
# build and run the app, passing parameters to it
dotnet run -- a b c
# app restarted, confirm parameters
Get-WmiObject Win32_Process | where name -eq your_app_name.exe | select commandline
```

@hughbe any thoughts on how else we can test this?

Comment (EatonZ) [reply src/System.Windows.Forms/src/System/Windows/Forms/Application.cs]: @RussKie It seems like we could use the scripts you made there, with 1 tweak: replace `=> System.Windows.Forms.Application.ExecutablePath;` with `=> System.Windows.Forms.Application.Restart();`, wrap that in a try/catch and report a failed test if it throws an exception.

Comment (RussKie) [reply src/System.Windows.Forms/src/System/Windows/Forms/Application.cs]: Yes. We still need to confirm all input arguments are being passed to the new instance.

**Comment by RussKie** [comment]
Sorry for the delay, it has slipped off my radar.
I've run tests locally, and confirmed the fix. Thank you 👍

**Pull request merged**